sopel/irc.py

```diff
@@ -214,9 +214,14 @@ def handle_connect(self):
                 try:
                     ssl.match_hostname(self.ssl.getpeercert(), self.config.core.host)
                 except ssl.CertificateError:
-                    stderr("Invalid certficate, hostname mismatch!")
-                    os.unlink(self.config.core.pid_file_path)
-                    os._exit(1)
+                    # try to recover if CNAMEs are the issue
+                    for cname in self._get_cnames(self.config.core.host):
+                        try:
+                            ssl.match_hostname(self.ssl.getpeercert(), cname)
+                        except ssl.CertificateError:
+                            stderr("Invalid certficate, hostname mismatch!")
+                            os.unlink(self.config.core.pid_file_path)
+                            os._exit(1)
```

Comment:
The way this is structured, it will exit if the first CNAME doesn't match and never try the others if they exist.

````
dgw@Kleiner $ python -i
Python 3.6.2 (default, Sep  7 2017, 13:16:50)
[GCC 6.3.0] on msys
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> listme = [0, 1, 2]
>>> for i in listme:
...     try:
...         raise IndexError
...     except IndexError:
...         print("Tried {}".format(i))
...         os._exit(1)
...
Tried 0
dgw@Kleiner $ 
````

Granted, having more than one CNAME is probably rare, but if the code _looks_ like it's trying to handle multiple CNAMEs, it _should_ handle them.

Reply:
You are right. I was quite tired when I hacked it together. I fixed it now.